iree/compiler/Conversion/LinalgToLLVM/PadLinalgWorkgroupTiles.cpp

```diff
@@ -0,0 +1,265 @@
+
+    std::tie(paddedMSize, paddingForM) = getWorkgroupPaddedTileSize(
+        problemSizeM, workgroupTileSizes[0], vectorTileSizes[0]);
+    std::tie(paddedNSize, paddingForN) = getWorkgroupPaddedTileSize(
+        problemSizeN, workgroupTileSizes[1], vectorTileSizes[1]);
+    std::tie(paddedKSize, paddingForK) =
+        getVectorPaddingSize(problemSizeK, vectorTileSizes[2]);
+
+    DEBUG_WITH_TYPE(DEBUG_TYPE, {
+      auto l1TileSizes = getTileSizes<TilingLevel::Level1Tiles>(matmulOp);
+      llvm::dbgs() << "Problem-size: "
+                   << "[" << problemSizeM << "," << problemSizeK << "]"
+                   << ", "
+                   << "[" << problemSizeK << "," << problemSizeN << "]\n";
+      llvm::dbgs() << "Workgroup-tile-sizes:"
+                   << "[" << workgroupTileSizes[0] << ", "
+                   << workgroupTileSizes[1] << "]\n";
+      llvm::dbgs() << "L1-tile-sizes:"
+                   << "[" << l1TileSizes[0] << ", " << l1TileSizes[1] << ","
+                   << l1TileSizes[2] << "]\n";
+      llvm::dbgs() << "Vector-tile-sizes:"
+                   << "[" << vectorTileSizes[0] << ", " << vectorTileSizes[1]
+                   << ", " << vectorTileSizes[2] << "]\n";
+      llvm::dbgs() << "LHS after padding:"
+                   << "[" << paddedMSize << "," << paddedKSize << "]\n";
+      llvm::dbgs() << "RHS after padding:"
+                   << "[" << paddedKSize << "," << problemSizeN << "]\n";
+    });
+
+    // No padding.
+    if (paddingForM == 0 && paddingForN == 0 && paddingForK == 0)
+      return failure();
+
+    auto getPaddedOperand = [&](Value operand, ArrayRef<int64_t> shape,
+                                ArrayRef<int64_t> highPadding) -> Value {
+      if (llvm::all_of(highPadding,
+                       [](int64_t val) -> bool { return val == 0; })) {
+        return operand;
+      }
+      auto elementType =
+          operand.getType().cast<RankedTensorType>().getElementType();
+      auto paddedType = RankedTensorType::get(shape, elementType);
+      auto paddingValue =
+          rewriter.create<ConstantOp>(loc, rewriter.getZeroAttr(elementType));
+      auto paddedOperand =
+          createPadTensorOpWithStaticPadding(rewriter, loc, paddedType, operand,
+                                             paddingValue, {0, 0}, highPadding);
+      return paddedOperand;
+    };
+
+    auto paddedLhs = getPaddedOperand(lhs, {paddedMSize, paddedKSize},
+                                      {paddingForM, paddingForK});
+
+    auto paddedrhs = getPaddedOperand(rhs, {paddedKSize, paddedNSize},
+                                      {paddingForK, paddingForN});
+
+    auto resultType = RankedTensorType::get(
+        {paddedMSize, paddedNSize},
+        result.getType().cast<RankedTensorType>().getElementType());
+
+    // Padding for K-dim only result doesn't change result size.
+    if (paddingForM == 0 && paddingForN == 0) {
+      auto paddedMatmulOp = rewriter.create<linalg::MatmulOp>(
```

Comment:
Wouldnt `clone` work (https://github.com/llvm/llvm-project/blob/5cb17728d19408725d4ced928ff3276dd4ffd1c9/mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.td#L1172).

Just changing the operands and the result type might help simplify all of this. Also `cloneAttributes` is good to avoid.